File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque
def solution(bridge_length, weight, truck_weights):
    answer = 0
    
    q = deque([0] * bridge_length)
    truck_weights = deque(truck_weights)
    weight_limit = 0
    
    while len(truck_weights) > 0:
        answer += 1
        weight_limit -= q.popleft()
        
        if weight_limit + truck_weights[0] <= weight:
            truck = truck_weights.popleft()
            weight_limit += truck
            q.append(truck)
        else:
            q.append(0)
        #print(answer, q)
    
    answer += bridge_length
    
    return answer

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    int total_weight = 0;
    queue<int> q;
    for (int i=0;i<bridge_length;i++){
        q.push(0);
    }
    queue<int> trucks;
    for (int i=0;i<truck_weights.size();i++){
        trucks.push(truck_weights[i]);
    }
    
    while (!trucks.empty()){
        answer++;
        total_weight -= q.front();
        q.pop();
        
        if (total_weight + trucks.front() <= weight){
            q.push(trucks.front());
            total_weight += trucks.front();
            trucks.pop();
        }
        else{
            q.push(0);
        }
    }
    
    answer += bridge_length;
    
    
    return answer;
}